Reorganize detrending

paigem opened the pull request:
- Replace scipy.signal.detrend with xarray.polyfit for 1-d fft detrend. If user asks for fft in only one dimension with linear detrend, xr.polyfit detrending is called.
- Reorganize `_apply_detrend()`. 
- Raise error if non-detrended dimensions do not have chunks of length 1

Worked together with @navidcy  

Closes #111, closes #113, closes #114

navidcy commented:
@roxyboy, are there any tests for detrending? If no, we should add some :)

navidcy commented:
This PR attempts to resolve #111

navidcy commented:
My comment above didn't seem to result in the issue being linked to the PR... Could somebody link #111?

roxyboy commented:
@navidcy The test for detrending is currently here: https://github.com/xgcm/xrft/blob/8b7e84eb414ed02fff059998d2396430417a1d29/xrft/tests/test_xrft.py#L80

navidcy commented:
how can I run the tests locally to make sure they pass OK?

roxyboy commented:
`pytest xrft -v` should do the trick. You need to have `pytest` installed in your local enviroment.

navidcy commented:
OK, @roxyboy allow me to have a look. I'll ping you.

paigem commented:
Thanks @roxyboy for the instructions on running tests locally. 

After running the test on our polyfit update, I am getting an error. I believe the error may be because polyfit accepts an xarray array, whereas the previous scipy.signal.detrend expected a numpy array.

**Specifics of the error:**
The line in `test_xrft.py` that brings up the error is: `da_prime = func(da.data, axes=[2]).compute()`. Here, `da.data` (i.e. *not* an xarray array) is being sent in as an input to `func()`, and understandably yields the error `AttributeError: 'Array' object has no attribute 'polyfit'`. We have written it such that, for ffts over 1d, the xarray array `da` should be sent to `func()`.

We have rewritten `_apply_detrend()` such that all 1d ffts (whether chunked or not) will be detrended via polyfit. The tests are currently written according to the previous logic: if the data is chunked at all, run `func()`.

Looking at it now, I think the 1d polyfit implementation we added should probably lie outside of `func()` and probably just in `_apply_detrend()`, but the same line from `test_xrft.py` will still raise an error since it calls `func()` directly with a numpy array and axes of length 1.

**My main question is:**
Can I change `test_xrft.py`, if others agree that it makes sense to do so? I'll have to look more closely to determine how to best update the test if so, but I just wanted to know if this is an option.

@navidcy what do you think?

navidcy commented:
Yes, we should change the tests to actually test what's implemented ;) Or write an additional test if needed.

(sorry I didn't have time to have a look yet...)

paigem commented:
7b4eafd: In this commit I moved the 1d polyfit implementation before any of the multidimensional screening in `func()`. This was done to help pass the tests that Navid updated.

navidcy commented on xrft/tests/test_xrft.py:
```diff
@@ -80,7 +80,7 @@ def numpy_detrend(da):
     N = 16
     x = np.arange(N+1)
-    y = np.arange(N-1)
+    y = np.arange(N-1) + 0.1*np.random.rand(N-1)
```
I added some non-triviality in the y axis (that is being detrended in the 1D FFTW tests. Otherwise, tests were failing simply because of slightly different implementation of detrending done via `scipy.signal.detrend()` and ours via `xarray.polyfit()`. We are simply removing the first order polynomial fit while `scipy.signal.detrend` does something a bit more elaborate.

navidcy commented:
@roxyboy, what do you think of these changes me and @paigem made?

roxyboy commented:
Thanks for the edits @paigem and @navidcy !!
Travis-ci seems to have passed but simply not showing up here... https://travis-ci.org/github/xgcm/xrft
I can't seem to figure out where the code coverage is decreasing with the PR... Any ideas?

navidcy commented:
@roxyboy we modified the tests/code a bit; I think now it should be good for you to have a careful look ;)

navidcy commented:
`test_detrend()` now also creates a 3D dummy data array to test that calling detrend without any axes still does what it's supposed to do

rabernat commented:
Thanks so much for this awesome work @paigem and @navidcy!

I will try to review by the end of the week.

> My comment above didn't seem to result in the issue being linked to the PR... Could somebody link #111?

This needs to be done in the original comment by the primary author (in this case @paigem). I edited that post to add the link.

paigem commented:
Thanks @rabernat, and sorry @navidcy - I didn't realize that was on me to link the issue here. 

I will also note that I haven't addressed my second comment in Issue #111 where I suggested we raise a warning if the user enters anything other than 'constant', 'linear', or 'None' that a warning should pop up. I'll look into that now, but it probably fits in a different PR at this point. 

What's the correct etiquette here - should I raise a new issue with that same comment, and then we can close that issue when the associated PR is merged?

navidcy commented:
The #111 issue's title only points to it's "part 1". I suggest edit the issue to only refer to part 1 and submit a separate issue for the warning?

navidcy commented:
#111's title only points to part 1. Also all the discussion in the issue is about part 1 (except a remark by @roxyboy that "a warning is a good idea". May I suggest split the issue into two: create a new issue about the warning and ~~strikeout~~ the part two from #111?

paigem commented:
Sounds good @navidcy. I will edit issue #111 and create a new one for the warning.

roxyboy commented:
Thanks again @navidcy and @paigem . I think this PR is ready to merge.

rabernat commented on xrft/tests/test_xrft.py:
```diff
@@ -99,16 +110,25 @@ def test_detrend():
+    # test detrending along all dimensions (total dimensions <4)
+    da_prime = func(da.data).compute()
+    npt.assert_allclose(da_prime,
+                        xrft.detrendn(d3d, axes=[0,1,2]))
```
I am trying to understand how these tests work, and I'm getting tied up in circles. This is not the fault of this PR (tests were already there when you started), but I figure we should use this opportunity to try to make things a little more transparent.

What we have here is something like

```python
func = xrft.detrend_wrap(xrft.detrendn)
da3d = xr.DataArray(d3d, ...)
da = da3d.chunk({'time': None})
da_prime = func(da.data).compute()
npt.assert_allclose(da_prime, xrft.detrendn(d3d, axes=[0,1,2]))
```

It feels to me that this test is a bit circular. In the end, we are testing that the results of `detrendn` equal the results of `detrendn`!

Can we think of a way to refactor this test so that the purpose of each part is a little more clear and transparent?

If not, that is fine, and we can move forward without such changes.

navidcy replied:
Well... indeed @rabernat! They seem to be tautological. This was a problem with the test from before.
I made some changes (de512cf) to try to make the tests more strict actually asserting that the `detrend()` gives the expected output. This created a bit of boiler-plate code in `test_detrend()` -- I'll try to gather that up.

rabernat commented on xrft/xrft.py:
```diff
@@ -116,25 +116,26 @@ def detrend_wrap(detrend_func):
-                raise ValueError('The axis along the detrending is upon '
+            for each_axis in axes:
+                if len(a.chunks[each_axis]) != 1:
+                    raise ValueError('The axis that is being detrended '
```
Is this still a limitation now that we are using xarray's  built-in `polyfit`? Can't we relax this constraint?

navidcy replied:
Well, I think it *is* if user asks to detrend *more than 1 dimensions*. In that case `xarray.polyfit()` cannot be invoked and we fall back to the previous `detrendn()` function.

@rabernat, the lines you are pointing to are only called when `len(axes)>1`; see https://github.com/xgcm/xrft/blob/78d90a21f6768330a79ef07657ed1e2810cda6e6/xrft/xrft.py#L122

Any suggestion how to make this cleaner?

navidcy replied:
hm... I guess the question is whether we can use `xarray.polyfit` to somehow replace:
https://github.com/xgcm/xrft/blob/78d90a21f6768330a79ef07657ed1e2810cda6e6/xrft/xrft.py#L74-L102

to tell you the truth though I'm not 100% sure what the above-mentioned lines do... I guess somehow they best-linear hyperplane is computed and subtracted... I feel that there should be a place for `polyfit()` to do the job here, no?

paigem replied:
I agree that we should look into doing multidimensional detrending using `polyfit()`. I also find this bit of code confusing and thought we could make a first step by implementing `polyfit()` for 1d, and then figure out more dimensions. Let's look into this as well @navidcy!

rabernat commented on xrft/xrft.py:
```diff
@@ -148,16 +149,15 @@ def func(a, axes=None):
+    if len(axis_num) == 1:
+        func = detrend_wrap(detrendn)
+        da = func(da,axes=axis_num)
+    elif da.chunks:
```
Do we really need this check and this wrapper now that we are using `polyfit`?

navidcy replied:
hm... probably you are right @rabernat this may be redundant
our thinking with @paigem (if I recall correctly) was to deal with all detrending in the inner-most part of this wrapper... We (or at least I) were a bit confused with the wrappers.

@paigem, let's chat about how we can make this better or cleaner (or hopefully both)

paigem replied:
Yes, @navidcy and I debated whether to put the polyfit detrending within the wrapper function or not. Happy to discuss this more @navidcy!

rabernat reviewed (commented):
Thanks a lot for this great PR!

In reviewing the PR, I felt a bit confused about all the layers of "wrappers" in the code. Some of these may be unnecessary now that we are using polyfit. It seems we have the opportunity to remove some code and simplify xrft, which is always a good thing.

We don't have to address these concerns right now. They are really more for @roxyboy, who is the architect of this feature. But I thought I would at least ask before we move forward.

navidcy commented:
perhaps it's a good opportunity to deal with #113 in this PR? (if #113 requires something to be dealt with that is)

rabernat commented:
> perhaps it's a good opportunity to deal with #113 in this PR? (if #113 requires something to be dealt with that is)

Yes, I would say, as long as your hands are deep into this code, this is a good chance to resolve #113.

rabernat commented on xrft/xrft.py:
```diff
@@ -116,25 +116,26 @@ def detrend_wrap(detrend_func):
+        elif len(axes) > 3:
+            raise ValueError("Detrending is only supported up to "
+                            "3 dimensions.")
+
```
I would recommend using the following functions here:

```python
def _detrend_dim(da, dim, deg=1):
    # detrend a single dimension
    p = da.polyfit(dim=dim, deg=deg)
    fit = xr.polyval(da.time, p.polyfit_coefficients)
    return da - fit

def _detrend(da, dims, deg=1):
    # detrend multiple dimensions in serial
    # only valid for linear case (deg=1)
    da_detrended = da
    for dim in dims:
        da_detrended = _detrend_dim(da, dim, deg=deg)
    return da_detrended
```

paigem replied:
Thanks for your suggestion @rabernat. Using `polyval()` is clearly the way to go - didn't realize that was an option!

For the detrending along multiple dimensions, it looks like you are suggesting to loop through each dimension and perform the 1d detrend on each dimension. My understanding is that, if we want to do, for example, a 2d detrend (e.g. if I am calculating a FFT over 2 dimensions) we will need to fit a plane to the 2d data and subtract that off. I don't think this is the same as doing a 1d linear fit in the 2 dimensions, but perhaps I'm misunderstanding something here.

paigem replied:
@navidcy and I briefly discussed trying to implement dask's `linalg.lstsq()` for more than 1d detrends, since I think `polyfit()` only works for 1d.

rabernat replied:
> My understanding is that, if we want to do, for example, a 2d detrend (e.g. if I am calculating a FFT over 2 dimensions) we will need to fit a plane to the 2d data and subtract that off. I don't think this is the same as doing a 1d linear fit in the 2 dimensions, but perhaps I'm misunderstanding something here.

I think you're absolutely right. 😊  Now I am remembering we dealt with the same question when @roxyboy first implemented this feature. He actually implemented up to 3D detrending using fancy linear algebra (see [code here](https://github.com/xgcm/xrft/blob/8b7e84eb414ed02fff059998d2396430417a1d29/xrft/xrft.py#L53)). However, that code does not handle dask data properly in all cases.

Xarray's polyfit actually uses dask.linalg.lstsq under the hood. It is called via this sequence:
- https://github.com/pydata/xarray/blob/2ce1cfc3e33f4ac7b420a51ac54f39dff1ce008c/xarray/core/dataset.py#L6095-L6097
- https://github.com/pydata/xarray/blob/2ce1cfc3e33f4ac7b420a51ac54f39dff1ce008c/xarray/core/duck_array_ops.py#L624-L629
- https://github.com/pydata/xarray/blob/2ce1cfc3e33f4ac7b420a51ac54f39dff1ce008c/xarray/core/dask_array_ops.py#L115-L141

However, it may be very hard to do the 2D case this way from within xrft.

In order to keep this PR from getting too complicated, I recommend we do something like this:

```python 
def detrend(data, dims, ...):
    # data is still a DataArray at this point
    if len(dims)==1:
        return new_detrend_1D_with_polyfit(data, dims[0])
    else:
        return legacy_detrend_function(data, dims)
```

navidcy replied:
I agree -- let's do this and open an issue for ≥2 -dimensional detrending

paigem replied:
Sounds good to me!

navidcy commented:
@paigem, do lines 
https://github.com/xgcm/xrft/blob/4bdb6fce5b5e4bf96e7b579ac6f5c87ba3cca447/xrft/xrft.py#L130-#L132
needed or should I delete them completely?
(this is #113)

paigem commented:
Newest PR reorganizes the detrending portion of the code. With @navidcy. Also removed several lines that I deemed redundant, but I might have gone a bit overboard so best if others take a look as well. Major changes:

- both constant and linear detrends are now done in `apply_detrend()` (note that the leading underscore has been removed, since this function is now exported so it can be tested)
- we now use `polyval()` in conjunction with `polyfit()` per @rabernat's suggestion above
- added a line about valid detrend arguments to address #114 
- removed warning in `dft()` for axes being detrended that have chunks not equal to 1, since this appears to be addressed by simply defining chunks of length 1 (currently being done within `apply_detrend()`

paigem commented:
> @paigem, do lines
> https://github.com/xgcm/xrft/blob/4bdb6fce5b5e4bf96e7b579ac6f5c87ba3cca447/xrft/xrft.py#L130-#L132
> needed or should I delete them completely?
> (this is #113)

Good point - I had just looked at this! I personally think they can be deleted.

navidcy commented:
@paigem, what about
https://github.com/xgcm/xrft/blob/4bdb6fce5b5e4bf96e7b579ac6f5c87ba3cca447/xrft/xrft.py#L123-#L126
shall I delete them (see #113)?

paigem commented:
> shall I delete them (see #113)?

@navidcy I was just looking at this, and I think we can go ahead and delete these lines, assuming the tests still pass.

navidcy commented on xrft/xrft.py:
```diff
@@ -15,7 +15,7 @@
 
 
-__all__ = ["detrendn", "detrend_wrap",
+__all__ = ["detrendn", "detrend_wrap","apply_detrend",
```
One remark: the `apply_detrend` method was upgraded from private to public
This was mostly done so that it could be tested. Perhaps there is an alternative way to use a private function in tests other than upgrading it to public?

If we decide to keep `apply_detrend` as public then it deserves an upgrade in its docstring! :)

rabernat replied:
I didn't think any of the detrend functions should be part of the public API. Although I can see how the high-level `apply_detrend` could be a useful standalone utility.

Private functions should start with an underscore.

Regardless, you can still import private functions from other modules in the tests, e.g. `from xrft.xrft import _detrendn`.

rabernat commented on xrft/xrft.py:
```diff
@@ -116,50 +116,52 @@ def detrend_wrap(detrend_func):
+        # For > 1d detrend, need all non-FFT axes to have chunks of length 1
+        for d in da.dims:
+            if d not in dim:
+                da = da.chunk({d:1})
```
```suggestion
                raise ValueError(f"Cannot apply detrending over dimension {d} because it is chunked. "
                                 f"Please rechunk your data first by calling `da.chunk({{d}: 1)`. ")
```

I would definitely not recommend silently and automatically rechunking the data. This could lead to big problems for large datasets. Instead, raise an error and ask the user to explicitly rechunk the data.

We would also need to make sure this scenario is covered by a test (e.g. `pytest.raises`).

paigem replied:
Sounds good. The current version of the code does this chunking, and to be honest I don't fully understand why it's needed (beyond the fact that it's for the 2d and 3d detrending via matrix algebra) so left it in. But I agree, we should raise an error instead of doing automatically.

paigem replied:
Clarification: these lines here actually chunk the data into chunks of length 1, so the error should state that dimensions not being detrended over must be chunked. I deleted a warning in the last PR that I will add back in to this effect.

rabernat reviewed (changes requested):
Thanks so much for continuing to iterate on this. I think we are really almost there!

navidcy commented:
I made `_apply_detrend` private and we can discuss whether `apply_detrend` and/or `apply_window` should be made private in a separate issue/PR

navidcy commented:
found a few cases that throw warning errors that were not covered in the tests and fixed
now codecov is happy :)

navidcy commented:
I'll stop.. I feel the `test_detrend()` is now sort of overkill :)

@rabernat, have a look at the changes we've made with @paigem at your convenience

rabernat reviewed (approved):
This looks great! Thanks for your patience digging so deep into this messy code. Really appreciate the hard work here!

navidcy commented:
@rabernat, perhaps a new release after this merge?
